Withdrawals as part of core and API transactions

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#1881 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- 74abe1ca1af0dd31ff90dfa95d8f718fc463c8d0
  :round_pushpin: **add withdrawals to API & core transaction data types**
  Still to be done: withdrawals must be discovered when processing blocks

- 0211c5c20b46fec3df6f143ba65eec545db94374
  :round_pushpin: **upgrade Byron & Jormungandr compatibility code with empty withdrawals**
  Withdrawals are only a thing in Shelley, so for this targets, we can safely replace them by empty lists

- 861a79250cb37e730b41b6db38033f290a4ac1fe
  :round_pushpin: **wire withdrawals in Shelley compatibility code**
  Here, we extract withdrawals directly from the TxBody, the conversion is straightforward

- c36f23a8e673bf32dc7e8554e5bf7fe939d39af5
  :round_pushpin: **include withdrawals where relevant in test and arbitrary generators**
  
- 9f6738ed03318f1d93e7bb2f91665dbe693ba73b
  :round_pushpin: **add withdrawals to the API Swagger specification**
  We represent them as a list and not a key:value map with stake address as key because clients typically don't know the stake address associated with their wallet, so it'd be quite impractical for them to access this data if returned as a map...

- e7fcedb274768d8514b787c746eb758997fa9ef9
  :round_pushpin: **add roundtrip tests for stake address and show that withdrawals are returned from the API**
  
- 00c2d697db0e104dea24fea105df78d6200d6151
  :round_pushpin: **Add stake address constraints to Jormungandr and Byron**
  
- 533b772fb7567f855afe6753ca39d51830504227
  :round_pushpin: **add 'minWithdrawal' query parameter to the swagger specification**
  
- 07a0f235103882d2ba24416e0a45b34976d8ff2a
  :round_pushpin: **extend transaction API to allow filtering by minimal withdrawals amount**
  This is a 'any' sort of filter, such that it returns any transaction where there's at least one withdrawal matching the criteria. The use-case is typically to ask for all transactions where there's a min withdrawal of 1, such that we get the entire withdrawal history

- 3dd7257a2c76ffe453092d7764effa562e3c5e8c
  :round_pushpin: **review integration and unit tests, add additional checks to cover withdrawals**



# Comments

<!-- Additional comments or screenshots to attach if any -->

:warning: Sorry, the PR spans over many files because I needed to add a new class constraints `DecodeStakeAddress` which goes side-by-side with the current `DecodeAddress`.

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
add withdrawals to API & core transaction data types

Still to be done: withdrawals must be discovered when processing blocks

KtorZ committed:
upgrade Byron & Jormungandr compatibility code with empty withdrawals

Withdrawals are only a thing in Shelley, so for this targets, we can safely replace them by empty lists

KtorZ committed:
wire withdrawals in Shelley compatibility code

Here, we extract withdrawals directly from the TxBody, the conversion is straightforward

KtorZ committed:
include withdrawals where relevant in test and arbitrary generators

KtorZ committed:
add withdrawals to the API Swagger specification

We represent them as a list and not a key:value map with stake address as key because clients typically don't know the stake address associated with their wallet, so it'd be quite impractical for them to access this data if returned as a map...

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -491,6 +463,30 @@ migrateManually tr defaultFieldValues =
+                _ <- Sqlite.step query
+                Sqlite.finalize query
+            ColumnPresent ->
+                traceWith tr $ MsgManualMigrationNotNeeded field
```
I originally started adding the withdrawal to the `TxMeta` table but I reverted and went for a new table instead since it behaves much more like current TxIn and TxOut. I however kept this refactor as I thought it wouldn't hurt :blush:

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:
```diff
@@ -125,6 +125,18 @@ TxOut
+    txWithdrawalAccount W.ChimericAccount   sql=account
+
+    Primary txWithdrawalTxId txWithdrawalAccount
+    deriving Show Generic
```
Interesting stuff actually here :point_up:

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -820,16 +820,21 @@ data Tx = Tx
+    , withdrawals
+        :: !(Map ChimericAccount Coin)
+        -- ^ Withdrawals (of funds from a registered reward account) embedded in
+        -- a transaction. The order does not matter.
```
And here :point_up_2:

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -816,6 +832,61 @@ toStakePoolDlgCert xpub (W.PoolId pid) =
+                Right a
+      where
+        msg = TextDecodingError
+            "Unable to decode stake-address: not a well-formed address."
```
I'll add a roundtrip test for these instances.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -911,9 +983,6 @@ instance Buildable addr => Buildable (ConnectionId addr) where
     build (LocalAddress p) = build p
 
-instance Buildable W.ChimericAccount where
-    build (W.ChimericAccount addr) = hexF addr
```
Found this little orphan :upside_down_face: ... No longer needed since it's define in the core package next to its relevant type declaration.

KtorZ commented on specifications/api/swagger.yaml:
```diff
@@ -460,6 +465,20 @@ x-transactionOutputs: &transactionOutputs
+      - amount
+    properties:
+      stake_address: *stakeAddress
+      amount: *amount
```
Here, I've chosen not to represent withdrawals as a key:value map in the API because it would otherwise be a key:value map where all keys are pretty much unknown by the client :no_mouth: ... wallet clients aren't typically aware of what their stake address is, so having something like:

```json
{ "stake_addr1sjck9mdmfyhzvjhydcjllgj9vjvl522w0573ncustrrr2rg7h9azg4cyqd36yyd48t5ut72hgld0fg2x": 42 }
```

Would make it quite hard for them to inspect the content of that value ^^"... Hence, in the API, I am turning the `Map` we have internally into a list of object with a `stake_address` and `amount` properties.

paweljakubas replied:
yeah, sound decision

paweljakubas replied:
Was thinking before how absent withdrawals are going to be represented:  either by 0-length array or by making `withdrawals` in `ApiTransaction` not required... Why the first option is better?

KtorZ replied:
It'll be a 0-length array.

paweljakubas commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -534,9 +539,15 @@ data ApiTransaction (n :: NetworkDiscriminant) = ApiTransaction
     , status :: !(ApiT TxStatus)
     } deriving (Eq, Generic, Show)
 
+data ApiWithdrawal n = ApiWithdrawal
```
:+1:

paweljakubas commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -660,15 +671,20 @@ fromShelleyTx
     , mapMaybe fromShelleyRegistrationCert (toList certs)
     )
 
+fromShelleyWdrl :: SL.Wdrl TPraosStandardCrypto -> Map W.ChimericAccount W.Coin
```
:+1:

paweljakubas commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -387,15 +387,15 @@ computeTxSize proxy action cs =
     dummyKeyHashRaw = BS.pack (replicate 28 0)
 
-    withdrawals = mkWithdrawals
+    wdrls = mkWithdrawals
```
frankly speaking like more previous naming, but understand the probable rationale (80-line styling)

KtorZ replied:
Yeah, I usually try to keep naming consistent with the surrounding context.

paweljakubas reviewed (approved):
LGTM! would be good not to forget `_decodeStakeAddress . _encodeStakeAddress` roundtrip. Also some changes are needed in `/build/cardano-wallet/lib/core/test/bench/db/Main.hs:408` but it is no surprise when changes in such a core place are made.

KtorZ committed:
add roundtrip tests for stake address and show that withdrawals are returned from the API

KtorZ committed:
Add stake address constraints to Jormungandr and Byron

KtorZ committed:
add 'minWithdrawal' query parameter to the swagger specification

KtorZ committed:
extend transaction API to allow filtering by minimal withdrawals amount

This is a 'any' sort of filter, such that it returns any transaction where there's at least one withdrawal matching the criteria. The use-case is typically to ask for all transactions where there's a min withdrawal of 1, such that we get the entire withdrawal history

KtorZ committed:
review integration and unit tests, add additional checks to cover withdrawals

Commit:
Regenerate nix
